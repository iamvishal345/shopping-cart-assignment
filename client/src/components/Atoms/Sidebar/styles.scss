@import "assets/scss/_breakpoints.scss";
.sidebar-container {
  min-width: 200px;
  width: 30%;
  background-color: #eeeeee;
  @include xm {
    width: 100%;
    background-color: #d10157;
  }
  .mobile-dropdown {
    display: none;
    @include xm {
      display: flex;
      justify-content: space-between;
      align-items: center;
      color: white;
      padding: 1rem;
      p {
        margin: 0;
      }
      button {
        background: none;
        border: none;
      }
    }
  }
  .menu-list {
    list-style: none;
    padding: 0 0 0 0.5rem;
    margin-top: 0;
    .menu-item {
      border-bottom: 1px solid #d1d1d1;

      button {
        border: none;
        width: 100%;
        color: #757575;
        font-size: 14px;
        padding: 0;
        padding: 0.6rem 1rem;
        text-align: start;
        &.active {
          color: #d10157;
        }
      }
    }
    @include xm {
      display: none;
      &.mobile-active {
        display: block;
        position: absolute;
        width: 100%;
        padding: 0;
      }
    }
  }
}
