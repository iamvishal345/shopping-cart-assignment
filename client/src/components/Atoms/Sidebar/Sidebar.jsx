import React, { useState } from "react";
import "./styles.scss";

export const Sidebar = ({ menuList, selectedCallback }) => {
  const [selectedMenu, setSelectedMenu] = useState({
    key: "all",
    name: "All Categories",
  });
  const [showMenu, setShowMenu] = useState(false);
  return (
    <div className="sidebar-container">
      <div className="mobile-dropdown">
        <p>{selectedMenu.name}</p>
        <button
          type="button"
          onClick={() => setShowMenu(!showMenu)}
          aria-label={`${showMenu ? "Hide Dropdown" : " Show Dropdown"}`}
        >
          <svg
            viewBox="0 0 140 140"
            width="24"
            height="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="m121.3,34.6c-1.6-1.6-4.2-1.6-5.8,0l-51,51.1-51.1-51.1c-1.6-1.6-4.2-1.6-5.8,0-1.6,1.6-1.6,4.2 0,5.8l53.9,53.9c0.8,0.8 1.8,1.2 2.9,1.2 1,0 2.1-0.4 2.9-1.2l53.9-53.9c1.7-1.6 1.7-4.2 0.1-5.8z"
                fill="white"
              />
            </g>
          </svg>
        </button>
      </div>
      <aside>
        <ul className={`menu-list ${showMenu ? "mobile-active" : ""}`}>
          <li className="menu-item">
            <button
              className={`${selectedMenu.key === "all" ? "active" : ""}`}
              type="button"
              onClick={() => {
                setSelectedMenu({
                  key: "all",
                  name: "All Categories",
                });
                selectedCallback("all");
              }}
            >
              All Categories
            </button>
          </li>
          {menuList.map((menu) => (
            <li className="menu-item" key={menu.id}>
              <button
                className={`${selectedMenu.key === menu.key ? "active" : ""}`}
                type="button"
                onClick={() => {
                  setSelectedMenu(menu);
                  selectedCallback(menu.id);
                }}
              >
                {menu.name}
              </button>
            </li>
          ))}
        </ul>
      </aside>
    </div>
  );
};
