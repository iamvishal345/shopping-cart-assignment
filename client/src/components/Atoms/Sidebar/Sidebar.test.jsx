import React from "react";
import { shallow } from "enzyme";
import { Sidebar } from "./Sidebar";

const menuList = [
  {
    name: "Beverages",
    key: "beverages",
    description:
      "Our beverage department will ensure your fridge is always fully stocked. Shop for soda, juice, beer and more. ",
    enabled: true,
    order: 3,
    imageUrl: "/static/images/category/beverages.png",
    id: "5b675e5e5936635728f9fc30",
  },
  {
    name: "Bakery Cakes and Dairy",
    key: "bakery-cakes-dairy",
    description:
      "The best cupcakes, cookies, cakes, pies, cheesecakes, fresh bread, biscotti, muffins, bagels, fresh coffee, milk and more.",
    enabled: true,
    order: 2,
    imageUrl: "/static/images/category/bakery.png",
    id: "5b6899123d1a866534f516de",
  },
  {
    name: "Beauty and Hygiene",
    key: "beauty-hygiene",
    description:
      "Buy beauty and personal care products online in India at best prices.",
    enabled: true,
    order: 4,
    imageUrl: "/static/images/category/beauty.png",
    id: "5b68994e3d1a866534f516df",
  },
];
const selectedCallback = jest.fn();
const wrapped = shallow(
  <Sidebar menuList={menuList} selectedCallback={selectedCallback} />
);
describe("Sidebar", () => {
  it("should match the snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });

  it("should render menuList + 1(All) no of menus", () => {
    expect(wrapped.find(".menu-item").length).toEqual(menuList.length + 1);
  });

  it("should select menu from sidebar on click of menu item and set its class as active", () => {
    //0th index is All categories
    wrapped.find(".menu-item").at(1).find("button").simulate("click", {});
    expect(wrapped.find("li").at(1).find("button").prop("className")).toEqual(
      "active"
    );
    expect(selectedCallback).toBeCalledWith(menuList[0].id);

    //for All
    wrapped.find(".menu-item").at(0).find("button").simulate("click", {});
    expect(wrapped.find("li").at(0).find("button").prop("className")).toEqual(
      "active"
    );
    expect(selectedCallback).toBeCalledWith("all");
  });
  it("should test mobile menu", () => {
    wrapped.find(".mobile-dropdown").find("button").simulate("click", {});
    expect(wrapped.find("ul").prop("className")).toEqual(
      "menu-list mobile-active"
    );
    expect(
      wrapped.find(".mobile-dropdown").find("button").prop("aria-label")
    ).toEqual("Hide Dropdown");
  });
});
