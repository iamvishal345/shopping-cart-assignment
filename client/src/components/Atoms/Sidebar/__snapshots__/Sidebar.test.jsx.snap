// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should match the snapshot 1`] = `
<div
  className="sidebar-container"
>
  <div
    className="mobile-dropdown"
  >
    <p>
      All Categories
    </p>
    <button
      aria-label=" Show Dropdown"
      onClick={[Function]}
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 140 140"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="m121.3,34.6c-1.6-1.6-4.2-1.6-5.8,0l-51,51.1-51.1-51.1c-1.6-1.6-4.2-1.6-5.8,0-1.6,1.6-1.6,4.2 0,5.8l53.9,53.9c0.8,0.8 1.8,1.2 2.9,1.2 1,0 2.1-0.4 2.9-1.2l53.9-53.9c1.7-1.6 1.7-4.2 0.1-5.8z"
            fill="white"
          />
        </g>
      </svg>
    </button>
  </div>
  <aside>
    <ul
      className="menu-list "
    >
      <li
        className="menu-item"
      >
        <button
          className="active"
          onClick={[Function]}
          type="button"
        >
          All Categories
        </button>
      </li>
      <li
        className="menu-item"
        key="5b675e5e5936635728f9fc30"
      >
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          Beverages
        </button>
      </li>
      <li
        className="menu-item"
        key="5b6899123d1a866534f516de"
      >
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          Bakery Cakes and Dairy
        </button>
      </li>
      <li
        className="menu-item"
        key="5b68994e3d1a866534f516df"
      >
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          Beauty and Hygiene
        </button>
      </li>
    </ul>
  </aside>
</div>
`;
